// @or-sdk/tags/src/types.ts
import { Token } from '@or-sdk/base';

export type TagsConfig = {
  /**
   * token
   */
  token: Token;

  /**
   * Url of OneReach service discovery api
   */
  discoveryUrl?: string;

  /**
   * Account ID for cross-account requests (super admin only)
   */
  accountId?: string;

  /**
   * Url of OneReach DataHubSvc api
   */
  dataHubSvcUrl?: string;
};

export interface Taggable<T> {
  addTags(source: T, tagNames: string[]): Promise<T>;
  removeTags(source: T, tagNames: string[]): Promise<T>;
}

export type Tag = {
  id?: string;
  data: {
    label: string;
    color?: string;
    description?: string | null;
  };
  isDeleted?: boolean;
  dateCreated?: number;
};

export type ListTagsParams = {
  query?: {
    [key: string]: unknown;
  };
  projection?: string[];
  group?: string[];
  sandbox?: boolean;
};

export type PaginationOptions = {
  limit?: number;
  offset?: number;
};

export type GetTagParams = {
  query?: {
    [key: string]: unknown;
  };
  projection?: string[];
  sandbox?: boolean;
};
