// @or-sdk/flows/src/types.ts
import { Token } from '@or-sdk/base';

export type FlowsConfig = {
  /**
   * token
   */
  token: Token;

  /**
   * Url of OneReach service discovery api
   */
  discoveryUrl?: string;

  /**
   * Account ID for cross-account requests (super admin only)
   */
  accountId?: string;

  /**
   * Url of OneReach DataHubSvc api
   */
  dataHubSvcUrl?: string;

  /**
   * Url of OneReach Deployer api
   */
  deployerUrl?: string;
};

export type ListFlowsParams = {
  includeDeleted?: boolean;
  includeExisting?: boolean;
  query?: {
    [key: string]: unknown;
  };
  projection?: string[];
  group?: string[];
  sandbox?: boolean;
};

export type ListDataOutsParams = {
  query?: {
    [key: string]: unknown;
  };
  projection?: string[];
  group?: string[];
  sandbox?: boolean;
};

export type PaginationOptions = {
  limit?: number;
  offset?: number;
};

export type GetFlowParams = {
  query?: {
    [key: string]: unknown;
  };
  projection?: string[];
  sandbox?: boolean;
};

export type DataOut = {
  flowId: string;
  botId: string;
  name: string;
  type: string;
  ttl: number;
  meta: string;
  outputExample: string;
  stepLabel: string;
  stepId: string;
};

export type StepTemplateToDelete = {
  id: string;
  [key: string]: unknown;
};

export type DownloadTemplateResult = {
  id: string;
  data: {
    label: string;
  };
};
