// @or-sdk/step-templates/dist/esm/utils/getNewStepTemplate.js
import { v4 as uuidv4 } from 'uuid';
function getNewStepTemplate(stepId) {
    return {
        id: stepId || uuidv4(),
        version: '0.0.1',
        cacheVersion: uuidv4(),
        label: '',
        icon: '',
        iconType: 'default',
        iconUrl: '',
        shape: 'circle',
        description: '',
        dateCreated: Date.now(),
        dateModified: Date.now(),
        publishedBy: '',
        template: `/* Your custom logic goes here. */
// This exits the step down the exit leg labeled \`next\`
return this.exitStep('next' /*, outputData */ );
//  This second argument is optional output data, which
//  will be stored in the step's merge field upon exit.`,
        categories: [],
        form: {
            style: '',
            template: '',
            code: '',
        },
        data: {
            exits: [],
        },
        modules: [],
        reporting: {
            step: {
                label: 'Step',
                type: 'step',
                enabled: true,
                tags: [],
            },
        },
        outputExample: null,
        formBuilder: {
            stepExits: [
                {
                    component: 'exitStatic',
                    data: {
                        id: 'next',
                        label: 'next',
                    },
                },
            ],
            stepInputs: [],
            hasDataOut: false,
            hasProcessError: false,
            outputExample: null,
            formTemplate: `<!-- Tread lightly here there be form elements from Design tab, you've been warned! -->
    <%= inputs ? inputs.join('\\n    ') : '' %>
<!-- Form elements from Design tab will end here -->`,
        },
        rawMode: false,
        hooks: null,
    };
}
export default getNewStepTemplate;
