// @or-sdk/bots/src/types.ts
import { Token } from '@or-sdk/base';

export type BotsConfig = {
  /**
   * token
   */
  token: Token;

  /**
   * Url of OneReach service discovery api
   */
  discoveryUrl?: string;

  /**
   * Account ID for cross-account requests (super admin only)
   */
  accountId?: string;

  /**
   * Url of OneReach DataHubSvc api
   */
  dataHubSvcUrl?: string;
};

export type Bot = {
  id?: string;
  data: {
    label: string;
    color: string | null;
    description: string;
    iconUrl: string;
    password: string;
    deploy: {
      logsTTL: number;
    };
  };
  dateCreated?: number;
  dateModified?: number;
  schemaVersion?: number;
  tags?: string[];
};

export type ListBotsParams = {
  includeDeleted?: boolean;
  includeExisting?: boolean;
  query?: {
    [key: string]: unknown;
  };
  projection?: string[];
  sandbox?: boolean;
  group?: string[];
};

export type PaginationOptions = {
  limit?: number;
  offset?: number;
};

export type GetBotsParams = {
  query?: {
    [key: string]: unknown;
  };
  projection?: string[];
  sandbox?: boolean;
};

export type DownloadTemplateResult = {
  id: string;
  data: {
    label: string;
  };
};
