// @or-sdk/step-templates/dist/cjs/utils/getNewStepTemplate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var uuid_1 = require("uuid");
function getNewStepTemplate(stepId) {
    return {
        id: stepId || (0, uuid_1.v4)(),
        version: '0.0.1',
        cacheVersion: (0, uuid_1.v4)(),
        label: '',
        icon: '',
        iconType: 'default',
        iconUrl: '',
        shape: 'circle',
        description: '',
        dateCreated: Date.now(),
        dateModified: Date.now(),
        publishedBy: '',
        template: "/* Your custom logic goes here. */\n// This exits the step down the exit leg labeled `next`\nreturn this.exitStep('next' /*, outputData */ );\n//  This second argument is optional output data, which\n//  will be stored in the step's merge field upon exit.",
        categories: [],
        form: {
            style: '',
            template: '',
            code: '',
        },
        data: {
            exits: [],
        },
        modules: [],
        reporting: {
            step: {
                label: 'Step',
                type: 'step',
                enabled: true,
                tags: [],
            },
        },
        outputExample: null,
        formBuilder: {
            stepExits: [
                {
                    component: 'exitStatic',
                    data: {
                        id: 'next',
                        label: 'next',
                    },
                },
            ],
            stepInputs: [],
            hasDataOut: false,
            hasProcessError: false,
            outputExample: null,
            formTemplate: "<!-- Tread lightly here there be form elements from Design tab, you've been warned! -->\n    <%= inputs ? inputs.join('\\n    ') : '' %>\n<!-- Form elements from Design tab will end here -->",
        },
        rawMode: false,
        hooks: null,
    };
}
exports.default = getNewStepTemplate;
